src/app/ref-datum-edit/ref-datum-edit.component.ts

```diff
@@ -0,0 +1,532 @@
+      return incorrect ? {incorrectValue: {value: control.value}} : null;
+    };
+  }
+
+  //convert deg min sec to dec degrees
+  azimuth = function (deg, min, sec) {
+    var azi = 0;
+    if (deg < 0) {
+        azi = -1.0 * deg + 1.0 * min / 60.0 + 1.0 * sec / 3600.0;
+        return (-1.0 * azi).toFixed(5);
+    }
+    else {
+        azi = 1.0 * deg + 1.0 * min / 60.0 + 1.0 * sec / 3600.0;
+        return (azi).toFixed(5);
+    }
+  }
+
+  //convert dec degrees to dms
+  deg_to_dms = function (deg) {
+    if (deg < 0) {
+        deg = deg.toString();
+
+        //longitude, remove the - sign
+        deg = deg.substring(1);
+    }
+    var d = Math.floor(deg);
+    var minfloat = (deg - d) * 60;
+    var m = Math.floor(minfloat);
+    var s = ((minfloat - m) * 60).toFixed(3);
+
+    return ("" + d + ":" + m + ":" + s);
+  }
+
+  changeLatLngUnit(event) {
+    this.setLatLngValidators();
+    if (event.value == "decdeg") {
+        this.latLngUnit = "decdeg";
+        //they clicked Dec Deg..
+        if ((this.form.controls.latdeg.value !== undefined && this.form.controls.latmin.value !== undefined && this.form.controls.latsec.value !== undefined) &&
+            (this.form.controls.londeg.value !== undefined && this.form.controls.lonmin.value !== undefined && this.form.controls.lonsec.value !== undefined)) {
+            //convert what's here for each lat and long
+            this.form.controls['latitude_dd'].setValue(this.azimuth(this.form.controls.latdeg.value, this.form.controls.latmin.value, this.form.controls.latsec.value));
+            this.form.controls['longitude_dd'].setValue(this.azimuth(this.form.controls.londeg.value, this.form.controls.lonmin.value, this.form.controls.lonsec.value));
+            this.incorrectDMS = false;
+        } else {
+            //show card telling them to populate all three (DMS) for conversion to work
+            this.incorrectDMS = true;
+        }
+    } else {
+        this.latLngUnit = "dms"
+        //they clicked dms (convert lat/long to dms)
+        if (this.form.controls.latitude_dd.value !== undefined && this.form.controls.longitude_dd.value !== undefined) {
+            var latDMS = (this.deg_to_dms(this.form.controls.latitude_dd.value)).toString();
+            var ladDMSarray = latDMS.split(':');
+            this.form.controls['latdeg'].setValue(ladDMSarray[0]);
+            this.form.controls['latmin'].setValue(ladDMSarray[1]);
+            this.form.controls['latsec'].setValue(ladDMSarray[2]);
+
+            var longDMS = this.deg_to_dms(this.form.controls.longitude_dd.value);
+            var longDMSarray = longDMS.split(':');
+            this.form.controls['londeg'].setValue((-(longDMSarray[0])).toString());
+            this.form.controls['lonmin'].setValue(longDMSarray[1]);
+            this.form.controls['lonsec'].setValue(longDMSarray[2]);
+        }
+    }
+  }
+
+  submit() {
+    this.form.markAllAsTouched();
```

Comment:
is the form being considered invalid unless each field is touched?

Reply:
The form shouldn't be invalid if fields are untouched.  The reason I added this here and in the site edit form is because some form values are updated with the code and not by the user, for example using radio buttons/checkboxes that aren't part of the form but that set some of the form values when checked.  So those form values may have been changed, but not touched if that makes sense

Reply:
ah i got you